### PR TITLE
Building chaos-scheduler

### PR DESCRIPTION
**Pull request by Sanjay1611**
This PR builds the scheduling logic and can be used as follows -

- specify `spec.schedule`
    ```
    type: "now"
    executionTime: "2020-05-11T20:30:00Z"
    startTime: "2020-05-12T05:47:00Z"
    endTime: "2020-05-12T05:52:00Z"
    minChaosInterval: "2m"   #format should be like "10m" or "2h" accordingly for minutes and hours
    instanceCount: "2"
    includedDays: 0-6
    ```
  1. `type` can be of 3 types `now`, `once` and `repeat`
  2. `executionTime` to be specified in case of `type=once`
  3. `startTime` to be specified in case of `type=repeat`
  4. `endTime` to be specified in case of `type=repeat`
  5. `minChaosInterval` to be specified in case of `type=repeat`
  6. `instanceCount` to be specified in case of `type=repeat`
  7. `includedDays` to be specified in case of `type=repeat`

- specify `spec.engineTemplateSpec`
  It will be the same as `ChaosEngineSpec` of `ChaosEngine` needed to be created by this 
  schedule

### CODE REVIEW

**Comment by Sanjay1611** [comment]
Example CR for `chaosschedule`
apiVersion: litmuschaos.io/v1alpha1
kind: ChaosSchedule
metadata:
  name: schedule-nginx
spec:
  chaosServiceAccount: chaos-scheduler
  schedule:
    type: "now"
    executionTime: "2020-05-11T20:30:00Z"
    startTime: "2020-05-12T05:47:00Z"
    endTime: "2020-05-12T05:52:00Z"
    minChaosInterval: "2m"   #format should be like "10m" or "2h" accordingly for minutes and hours
    instanceCount: "2"
    includedDays: 0-6
    random: false
  engineTemplateSpec:
    jobCleanUpPolicy: "retain"
    engineState: "active"
    auxiliaryAppInfo: ""
    appinfo:
      appns: default
      applabel: "app=nginx"
      appkind: deployment
    chaosServiceAccount: litmus
    monitoring: false
    experiments:
    - name: pod-delete

**Comment by chandankumar4** [comment]
Add `_output` dir in .gitignore

**Comment by Sanjay1611** [comment]
> Add `_output` dir in .gitignore

Added

**Comment by chandankumar4** [comment]
> > Add `_output` dir in .gitignore
> 
> Added

Remove this file as well: build/_output/bin/chaos-scheduler

File: deploy/chaos-scheduler.yaml [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  template:
+    metadata:
+      labels:
+        name: chaos-scheduler
+    spec:
+      serviceAccountName: chaos-scheduler
+      containers:
+        - name: chaos-scheduler
+          # Replace this with the built image name
+          image: REPLACE_IMAGE
```
Comment (chandankumar4) [inline deploy/chaos-scheduler.yaml:19]:
```suggestion
          image: litmuschaos/chaos-scheduler:latest
```

File: hack/update-codegen.sh [hunk 2]
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+${GOPATH}/src/k8s.io/code-generator/generate-groups.sh all \
+  github.com/litmuschaos/chaos-scheduler/pkg/client github.com/litmuschaos/chaos-scheduler/pkg/apis \
+  litmuschaos:v1alpha1
```
Comment (chandankumar4) [inline hack/update-codegen.sh:9]: Add these inside the makefile and remove this file

File: pkg/controller/chaosscheduler/chaosscheduler_controller.go [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+		// Don't do anything other than updating status.
+		return reconcile.Result{}, nil
+	}
+
+	cronString, duration, err := scheduleRepeat(cs)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// var times []time.Time
```
Comment (chandankumar4) [inline pkg/controller/chaosscheduler/chaosscheduler_controller.go:231]: remove this

File: pkg/controller/chaosscheduler/chaosscheduler_controller.go [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+	case "", schedulerV1.StateActive:
+		{
+			return schedulerReconcile.reconcileForCreationAndRunning(scheduler)
+		}
+	// case schedulerV1.StateStopped:
+	// 	{
+	// 		if !checkScheduleStatus(scheduler, schedulerV1.StatusStopped) {
+	// 			return r.reconcileForDelete(scheduler, request)
+	// 		}
+	// 	}
```
Comment (chandankumar4) [inline pkg/controller/chaosscheduler/chaosscheduler_controller.go:135]: is this not required?

File: pkg/controller/chaosscheduler/chaosscheduler_controller.go [hunk 5]
```diff
@@ -0,0 +1,336 @@
...
+	cronString, duration, err := scheduleRepeat(cs)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// var times []time.Time
+	scheduledTime, errNew := getRecentUnmetScheduleTime(cs, cronString)
+	if errNew != nil {
+		schedulerReconcile.r.recorder.Eventf(cs.Instance, corev1.EventTypeWarning, "FailedNeedsStart", "Cannot determine if engine needs to be started: %v", errNew)
+		// klog.Errorf("Cannot determine if %s needs to be started: %v", nameForLog, err)
```
Comment (chandankumar4) [inline pkg/controller/chaosscheduler/chaosscheduler_controller.go:235]: remove this

File: pkg/controller/chaosscheduler/utils.go [hunk 6]
```diff
@@ -0,0 +1,218 @@
+package chaosscheduler
```
Comment (chandankumar4) [inline pkg/controller/chaosscheduler/utils.go:1]: change this file name

File: pkg/controller/chaosscheduler/chaosscheduler_controller.go (hunk 4 above)
Comment (Sanjay1611) [reply pkg/controller/chaosscheduler/chaosscheduler_controller.go:135]: May be needed in the near future.... Not confirm whether we should have a `stop` state for the scheduler .. If yes then what will be its function

File: pkg/controller/chaosscheduler/utils.go (hunk 6 above)
Comment (Sanjay1611) [reply pkg/controller/chaosscheduler/utils.go:1]: As this file contains general miscellaneous functions.. I think this is the best suitable file name.

Comment (chandankumar4) [reply pkg/controller/chaosscheduler/utils.go:1]: PTAL: https://github.com/litmuschaos/litmus/blob/cfd91c26ca1ca668a40d2bc85fdbc23f576de8df/contribute/developer_guide/STYLE_GUIDE.md#package-naming

File: pkg/controller/chaosscheduler/chaosscheduler_controller.go (hunk 4 above)
Comment (Sanjay1611) [reply pkg/controller/chaosscheduler/chaosscheduler_controller.go]: Removed it for now

File: pkg/controller/chaosscheduler/utils.go (hunk 6 above)
Comment (Sanjay1611) [reply pkg/controller/chaosscheduler/utils.go]: Changed

**Comment by Sanjay1611** [comment]
I have added the circle-ci config file too. Do check it once

File: README.md [hunk 7]
```diff
@@ -1,2 +1,175 @@
...
-Schedule your Chaos!! 
+
+# Litmus chaos-scheduler for injecting chaos experiments on Kubernetes
+  
+Litmus chaos scheduler is used by Kubernetes application developers and SREs to inject chaos into the applications 
+and Kubernetes infrastructure in a managed fashion. Its objective is to make the process of validation and 
+hardening of application workloads on Kubernetes easy by automating the execution of chaos experiments. A sample chaos 
+injection workflow could be as simple as:
+
+- Install the Litmus infrastructure components (RBAC, CRDs), the scheduler & Experiment custom resource bundles via helm charts
```
Comment (ksatchit) [inline README.md:9]: operator also

File: README.md [hunk 8]
```diff
@@ -1,2 +1,175 @@
...
+## What is a chaos scheduler and how is it built?
+
+The Chaos scheduler is a Kubernetes scheduler, which are nothing but custom-controllers with direct access to Kubernetes API
+that can manage the lifecycle of certain resources or applications, while always trying to ensure the resource is in the "desired
+state". The logic that ensures this is commonly called "reconcile" function.
+
+The Chaos scheduler is built using the popular [Operator-SDK](https://github.com/operator-framework/operator-sdk/) framework, 
+which provides bootstrap support for new scheduler projects, allowing teams to focus on business/operational logic. 
+
+The Litmus Chaos scheduler helps reconcile the state of the ChaosEngine, a custom resource that holds the chaos intent
```
Comment (ksatchit) [inline README.md:30]: reconciles ChaosSchedule not ChaosEngine

File: README.md [hunk 9]
```diff
@@ -1,2 +1,175 @@
...
+The Chaos scheduler is a Kubernetes scheduler, which are nothing but custom-controllers with direct access to Kubernetes API
+that can manage the lifecycle of certain resources or applications, while always trying to ensure the resource is in the "desired
+state". The logic that ensures this is commonly called "reconcile" function.
+
+The Chaos scheduler is built using the popular [Operator-SDK](https://github.com/operator-framework/operator-sdk/) framework, 
+which provides bootstrap support for new scheduler projects, allowing teams to focus on business/operational logic. 
+
+The Litmus Chaos scheduler helps reconcile the state of the ChaosEngine, a custom resource that holds the chaos intent 
+specified by a developer/devops engineer against a particular stateless/stateful Kubernetes deployment. The scheduler performs
+specific actions upon CRUD of the ChaosEngine, its primary resource. The scheduler also defines secondary resources (the engine
```
Comment (ksatchit) [inline README.md:32]: ChaosSchedule

Comment (ksatchit) [reply README.md:32]: Secondary resource is ChaosEngine CR here not the runner pod nad monitor service.

File: README.md [hunk 10]
```diff
@@ -1,2 +1,175 @@
...
+
+- Application Data (namespace, labels, kind)
+- List of Chaos Experiments to be executed
+- Attributes of the experiments, such as, rank/priority 
+- Execution Schedule for the batch run of the experiments
+
+The ChaosEngine is the referenced as the owner of the secondary (reconcile) resource with Kubernetes deletePropagation 
+ensuring these also are removed upon deletion of the ChaosEngine CR.
+
+Here is a sample ChaosEngineSpec for reference:
```
Comment (ksatchit) [inline README.md:47]: The section on chaosengine and charts etc.., are not needed here. Just refer / provide link to the docs and operator readme. In this README the focus is purely on schedule.

File: README.md [hunk 11]
```diff
@@ -1,2 +1,175 @@
...
+    experiments:
+      - name: pod-delete 
+        spec:
+          rank: 
+      - name: container-kill
+        spec:
+          rank:  
+  ```
+
+## What is a litmus chaos chart and how can I use it?
```
Comment (ksatchit) [inline README.md:67]: same as previous comment on engine description.

File: README.md [hunk 12]
```diff
@@ -1,2 +1,175 @@
...
+      value: null
+    - name: TOTAL_CHAOS_DURATION
+      value: 15
+    - name: CHAOS_INTERVAL
+      value: 5
+    - name: LIB
+      value: ""
+```
+
+## What are the steps to get started?
```
Comment (ksatchit) [inline README.md:113]: Update these steps based on schedule steps

File: build/bin/entrypoint [hunk 13]
```diff
@@ -0,0 +1,12 @@
+#!/bin/sh -e
```
Comment (ksatchit) [inline build/bin/entrypoint:1]: can you explain these scripts?

File: pkg/controller/chaosscheduler/createEngineForNowAndOnce.go [hunk 14]
```diff
@@ -0,0 +1,68 @@
+package chaosscheduler
```
Comment (chandankumar4) [inline pkg/controller/chaosscheduler/createEngineForNowAndOnce.go:1]: These package naming we can use like below:
pkg/controller/chaosscheduler/chaosengine/now.go
pkg/controller/chaosscheduler/chaosengine/once.go
pkg/controller/chaosscheduler/chaosengine/repeat.go


Good for now we can do in the upcoming PR also

**Pull request merged**